core/src/main/scala/org/apache/spark/sql/delta/commands/DeletionVectorUtils.scala

```diff
@@ -82,6 +83,26 @@ trait DeletionVectorUtils {
     protocol.isFeatureSupported(DeletionVectorsTableFeature) &&
       metadata.format.provider == "parquet" // DVs are only supported on parquet tables.
   }
+
+  /**
+   * Utility method that checks the table has no Deletion Vectors enabled. Deletion vectors
+   * are supported only in read-mode for now. Any updates to tables with deletion vectors
+   * feature are disabled until we add support.
+   */
+  def assertNoDeletionVectors(spark: SparkSession, metadata: Metadata, protocol: Protocol): Unit = {
```

Comment:
@xupefei

Reply:
@larsk-db @xupefei Please review.

Reply:
you mean `assertDeletionVectorsNotWritable`?

Reply:
No, I mean `NotReadable`. I.e. the table feature is not set. That is what it checks.
The matrix of states for DVs is basically:

| | `reader/writerFeatures` does not contain `deletionVectors` | `reader/writerFeatures` contains `deletionVectors` |
| - | ----------------------------------------------------------|----------------------------------------------------|
| `delta.enableDeletionVectors = 'false'` (or absent) | DVs not readable | DVs readable (but not writable) |
| `delta.enableDeletionVectors = 'true'` | illegal state* | DV's readable and writable |

* I'm still not convinced we should interpret this as "DVs readable+writable" like @xupefei suggested. Imo it should throw an illegal state exception and make people fix their table by either removing the property or adding the features, whichever was the desired intent. It seems to hard to make a right guess here.

Reply:
The `illegal state` case won't happen at this stage. During Delta log creation we do assert on metadata <> protocol is in sync: https://github.com/delta-io/delta/blob/1c18c1d972e37d314711b3a485e6fb7c98fce96d/core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala#L98.

Reply:
We do guard this case in Snapshot construction: https://github.com/delta-io/delta/blob/1c18c1d972e37d314711b3a485e6fb7c98fce96d/core/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala#L96-L100.

The valid case is `reader/writerFeatures contains deletionVectors` + `delta.enableDeletionVectors = 'false' (or absent)` but not the other way around.

Reply:
But that case we don't have to check `DeletionVectorsTableFeature.metadataRequiresFeatureToBeEnabled(metadata, spark)` as you suggested below.

Reply:
I checked the call sites of `assertNoDeletionVectors`... The one in Tnx that asserts on `newMetadataTmp` should be preserved to avoid creating an ill-formed table during write (the above `assertTableFeaturesMatchMetadata` is only called during read). For the other two, it's not required to check `metadataRequiresFeatureToBeEnabled`.

Reply:
Then let's separate these into two different methods with names that make it clear what each of them does. (Or maybe one method for `...NotReadable` and then the `metadataRequiresFeatureToBeEnabled ` check directly in the txn code with a comment why it's necessary.